File: 01-Programming-Languages/Swift/Data-Structures/Graph/graph_adjacency_list.swift
/*
Problem: Graph Adjacency List
Description: Implement a graph using an adjacency list with
           addVertex, addEdge, removeEdge, display, and DFS.

Approach:
- Store graph as dictionary mapping vertex to list of neighbors
- Supports undirected graph operations
- DFS uses recursion with visited set

Time Complexity: O(V + E) for traversal
Space Complexity: O(V + E)

Example:
Input: add edges 0-1, 0-2, 1-2, 1-3
Output: DFS from 0: 0, 1, 2, 3
*/

class Graph {
    private var adjList = [Int: [Int]]()

    func addVertex(_ v: Int) {
        if adjList[v] == nil { adjList[v] = [] }
    }

    func addEdge(_ u: Int, _ v: Int) {
        addVertex(u); addVertex(v)
        adjList[u]?.append(v)
        adjList[v]?.append(u)
    }

    func removeEdge(_ u: Int, _ v: Int) {
        adjList[u]?.removeAll { $0 == v }
        adjList[v]?.removeAll { $0 == u }
    }

    func display() {
        for (vertex, neighbors) in adjList.sorted(by: { $0.key < $1.key }) {
            print("\(vertex): \(neighbors.sorted())")
        }
    }

    func dfs(from start: Int) -> [Int] {
        var visited = Set<Int>()
        var result = [Int]()
        dfsHelper(start, &visited, &result)
        return result
    }

    private func dfsHelper(_ v: Int, _ visited: inout Set<Int>, _ result: inout [Int]) {
        visited.insert(v)
        result.append(v)
        for neighbor in adjList[v] ?? [] where !visited.contains(neighbor) {
            dfsHelper(neighbor, &visited, &result)
        }
    }
}

let g = Graph()
g.addEdge(0, 1)
g.addEdge(0, 2)
g.addEdge(1, 2)
g.addEdge(1, 3)
print("Graph adjacency list:")
g.display()
print("DFS from 0: \(g.dfs(from: 0))")

File: 01-Programming-Languages/Dart/Data-Structures/Graph/graph_adjacency_list.dart
/*
Problem: Graph Adjacency List
Description: Implement a graph using an adjacency list with
           addVertex, addEdge, removeEdge, display, and DFS.

Approach:
- Store graph as map mapping vertex to list of neighbors
- Supports undirected graph operations
- DFS uses recursion with visited set

Time Complexity: O(V + E) for traversal
Space Complexity: O(V + E)

Example:
Input: add edges 0-1, 0-2, 1-2, 1-3
Output: DFS from 0: 0, 1, 2, 3
*/

class Graph {
  Map<int, List<int>> _adjList = {};

  void addVertex(int v) {
    _adjList.putIfAbsent(v, () => []);
  }

  void addEdge(int u, int v) {
    addVertex(u);
    addVertex(v);
    _adjList[u]!.add(v);
    _adjList[v]!.add(u);
  }

  void removeEdge(int u, int v) {
    _adjList[u]?.remove(v);
    _adjList[v]?.remove(u);
  }

  void display() {
    var sortedKeys = _adjList.keys.toList()..sort();
    for (int key in sortedKeys) {
      var neighbors = List<int>.from(_adjList[key]!)..sort();
      print('$key: $neighbors');
    }
  }

  List<int> dfs(int start) {
    Set<int> visited = {};
    List<int> result = [];
    _dfsHelper(start, visited, result);
    return result;
  }

  void _dfsHelper(int v, Set<int> visited, List<int> result) {
    visited.add(v);
    result.add(v);
    for (int neighbor in (_adjList[v] ?? [])) {
      if (!visited.contains(neighbor)) {
        _dfsHelper(neighbor, visited, result);
      }
    }
  }
}

void main() {
  Graph g = Graph();
  g.addEdge(0, 1);
  g.addEdge(0, 2);
  g.addEdge(1, 2);
  g.addEdge(1, 3);
  print('Graph adjacency list:');
  g.display();
  print('DFS from 0: ${g.dfs(0)}');
}
